### PR TITLE
New database api implementation

### PR DESCRIPTION
**Pull request by Eabramov84**

### CODE REVIEW

File: database/database.go [hunk 1]
```diff
@@ -874,3 +773,19 @@ func getInstancesFromQueryResult(rows *sql.Rows) (instances []launcher.InstanceI
 
 	return instances, nil
 }
+
+func getServicesFromFromQueryResult(rows *sql.Rows) (services []servicemanager.ServiceInfo, err error) {
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply database/database.go:808]: added

**Review by MykolaSolyanko — APPROVED** [review]

File: database/database_internal_test.go [hunk 2]
```diff
@@ -415,6 +416,170 @@ func TestLayers(t *testing.T) {
 	}
 }
 
+func TestInstances(t *testing.T) {
+	var (
+		subjectInstances, serviceInstances, running, allInstances []launcher.InstanceInfo
+		testServiceID                                             = "testService"
```
Comment (al1img) [inline database/database_internal_test.go:423]: make const

File: database/database_internal_test.go [hunk 3]
```diff
@@ -415,6 +416,170 @@ func TestLayers(t *testing.T) {
...
+	testInstanceInfo := launcher.InstanceInfo{InstanceIdent: cloudprotocol.InstanceIdent{
+		ServiceID: testServiceID,
+		SubjectID: testSubjectID, Instance: 42,
+	}, AosVersion: 42, UnitSubject: true, UID: 42}
+
+	testInstanceID := fmt.Sprintf("%s_%s_%s", testInstanceInfo.ServiceID, testInstanceInfo.SubjectID, strconv.Itoa(42))
+
+	testInstanceInfo.InstanceID = testInstanceID
+
+	if err := db.AddInstance(testInstanceInfo); err != nil {
```
Comment (al1img) [inline database/database_internal_test.go:523]: Move negative tests into separate test or add comments what is tested.

File: database/database.go [hunk 4]
```diff
@@ -337,6 +338,45 @@ func (db *Database) GetJournalCursor() (cursor string, err error) {
...
+	}
+
+	if err = json.Unmarshal([]byte(textEnvVars), &vars); err != nil {
+		return vars, aoserrors.Wrap(err)
+	}
+
+	return vars, nil
+}
+
+// SetOverrideEnvVars update override env vars.
```
Comment (al1img) [inline database/database.go:281]: update -> sets

File: database/database.go [hunk 5]
```diff
@@ -547,6 +548,62 @@ func (db *Database) GetServiceInstances(serviceID string) (instances []launcher.
...
+	}
+
+	if err != nil {
+		return info, aoserrors.Wrap(err)
+	}
+
+	return info, nil
+}
+
+// AddStorageStateInfo add storage and state info with instance ID.
```
Comment (al1img) [inline database/database.go:502]: add -> adds

File: database/database.go [hunk 6]
```diff
@@ -547,6 +548,62 @@ func (db *Database) GetServiceInstances(serviceID string) (instances []launcher.
...
+func (db *Database) AddStorageStateInfo(instanceID string, info storagestate.StorageStateInstanceInfo) error {
+	return db.executeQuery("INSERT INTO storagestate values(?, ?, ?, ?)",
+		instanceID, info.StorageQuota, info.StateQuota, info.StateChecksum)
+}
+
+func (db *Database) SetStorageStateQuotasByID(instanceID string, storageQuota, stateQuota uint64) (err error) {
```
Comment (al1img) [inline database/database.go:508]: Add comment.

File: database/database_internal_test.go [hunk 7]
```diff
@@ -622,6 +623,85 @@ func TestEnvVars(t *testing.T) {
...
+		t.Fatalf("Can't get state storage info: %v", err)
+	}
+
+	if !reflect.DeepEqual(info, testStateStorageInfo) {
+		t.Error("Update state storage info from database doesn't match expected one")
+	}
+
+	if err := db.SetStorageStateQuotasByID(
+		"noID", newStorageQuota, newStateQuota); !errors.Is(err, storagestate.ErrNotExist) {
+		t.Errorf("Should be entry does not exist")
```
Comment (al1img) [inline database/database_internal_test.go:739]: Should be:

File: database/database.go [hunk 8]
```diff
@@ -447,6 +447,37 @@ func (db *Database) GetServiceInstances(serviceID string) (instances []launcher.
...
+// GetInstanceIDs returns instance ids by filter.
+func (db *Database) GetInstanceIDs(filter cloudprotocol.InstanceFilter) (instances []string, err error) {
+	var subjectFiler, instanceFiler string
+
+	if filter.SubjectID != nil {
+		subjectFiler = fmt.Sprintf(" AND subjectID = \"%s\"", *filter.SubjectID)
+	}
+
+	if filter.Instance != nil {
+		instanceFiler = fmt.Sprintf(" AND instance = %v", *filter.Instance)
```
Comment (al1img) [inline database/database.go:459]: %d?

File: database/database.go [hunk 9]
```diff
@@ -447,6 +447,37 @@ func (db *Database) GetServiceInstances(serviceID string) (instances []launcher.
...
+	}
+
```
Comment (al1img) [inline database/database.go:461]: You can add  `serviceFilter` as well.

File: database/database.go [hunk 10]
```diff
@@ -447,6 +447,37 @@ func (db *Database) GetServiceInstances(serviceID string) (instances []launcher.
 	return getInstancesFromQueryResult(rows)
```
Comment (al1img) [inline database/database.go:447]: Fix commit message.

File: database/database.go [hunk 11]
```diff
@@ -408,6 +409,104 @@ func (db *Database) GetLayerInfoByDigest(digest string) (layer layermanager.Laye
...
+		instance.AosVersion, instance.UnitSubject, instance.Running, instance.UID)
+}
+
+// UpdateInstance update instance information in db.
+func (db *Database) UpdateInstance(instance launcher.InstanceInfo) (err error) {
+	if err = db.executeQuery(
+		`UPDATE instances SET serviceID = ?, subjectID = ?, instance = ?, aosVersion = ?, unitSubject = ?, running = ?,
+		 uid = ? WHERE  instanceID = ?`,
+		instance.ServiceID, instance.SubjectID, instance.Instance, instance.AosVersion, instance.UnitSubject,
+		instance.Running, instance.UID, instance.InstanceID); errors.Is(err, ErrNotExist) {
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply database/database.go:365]: will do as separate commit

File: database/database_internal_test.go [hunk 12]
```diff
@@ -415,6 +416,170 @@ func TestLayers(t *testing.T) {
...
+		testSubjectID                                             = "testSubject"
+	)
+
+	for i := 0; i < 5; i++ {
+		subjectInstance := launcher.InstanceInfo{InstanceIdent: cloudprotocol.InstanceIdent{
+			ServiceID: "someServiceID" + strconv.Itoa(i),
+			SubjectID: testSubjectID, Instance: uint64(i),
+		}, AosVersion: uint64(i), UnitSubject: true, UID: i + 100}
+
+		subjectInstance.InstanceID = fmt.Sprintf("%s_%s_%s", subjectInstance.ServiceID, subjectInstance.SubjectID,
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply database/database_internal_test.go:433]: it is already uniq one

File: database/database.go [hunk 13]
```diff
@@ -408,6 +409,104 @@ func (db *Database) GetLayerInfoByDigest(digest string) (layer layermanager.Laye
...
+	}
+
+	return err
+}
+
+// GetInstanceByIdent returns instance information by serviceID + subjectID + instance.
+func (db *Database) GetInstanceByIdent(instanceIdent cloudprotocol.InstanceIdent) (
+	instance launcher.InstanceInfo, err error,
+) {
+	stmt, err := db.sql.Prepare("SELECT * FROM instances WHERE serviceID = ? AND subjectID = ? AND instance = ?")
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply database/database.go:386]: done

File: database/database_internal_test.go (hunk 3 above)
Comment (Eabramov84) [reply database/database_internal_test.go:523]: added comment

File: database/database.go (hunk 9 above)
Comment (Eabramov84) [reply database/database.go:461]: added as ServiceOD to param

File: database/database_internal_test.go [hunk 14]
```diff
@@ -415,6 +417,167 @@ func TestLayers(t *testing.T) {
...
+	}
+
+	testInstanceInfo := launcher.InstanceInfo{InstanceIdent: cloudprotocol.InstanceIdent{
+		ServiceID: testServiceID,
+		SubjectID: testSubjectID, Instance: 42,
+	}, AosVersion: 42, UnitSubject: true, UID: 42}
+
+	testInstanceID := fmt.Sprintf("%s_%s_%s", testInstanceInfo.ServiceID, testInstanceInfo.SubjectID, strconv.Itoa(42))
+
+	testInstanceInfo.InstanceID = testInstanceID
```
Comment (al1img) [inline database/database_internal_test.go:517]: Generate instance ID at initialization, use the same mechanism as before.

**Review by al1img — APPROVED** [review]

**Pull request merged**